Coupons: Enable Save button on Edit Coupon screen only when there are changes made

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6832  
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds logic to check for changes on the Edit Coupon screen, so that the Save button is only enabled if there are any changes made to the coupon details.

This also updates the way `populatedCoupon` is created - previously we're using `copy` to update values for the existing coupon. This method ignores nil values, so we cannot remove the expiry date for the coupon by sending nil to the method. The solution is to create a new Coupon manually instead.


### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Make sure that you have Coupon Management enabled in Settings > Experimental Features.
- Navigate to Menu > Coupons > select a coupon > select the ellipsis button and choose Edit Coupon.
- Notice that the Save button is disabled initially. Editing any of the fields should enable the button.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/5533851/168049177-7f7f1802-f76d-415f-bcf4-295159c88d09.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Check changes in coupon details

itsmeichigo committed:
Show the save button only if there's any change made

itsmeichigo committed:
Only enable/disable save button instead of showing/hiding

itsmeichigo committed:
Update populatedCoupon to not use copy

itsmeichigo committed:
Add some unit tests for hasChangesMade

itsmeichigo committed:
More unit tests for hasChangesMade

itsmeichigo committed:
Revert change to usageLimit check

pmusolino committed:
fix: remove expiration date when coupon is expired

pmusolino reviewed (commented):
I just noticed that with some coupons the button is always enabled, and for other coupons no.
I believe that we are changing some information under the hood, and we are not taking in consideration that changes 🤔 WDYT?

https://user-images.githubusercontent.com/495617/168105987-742ecf1d-92e7-41c9-af16-83e38f18e269.mp4

itsmeichigo committed:
Move the clearing of time from date to the date picker

itsmeichigo commented:
Hi @pmusolino - the issue seems to be caused by the trimming of time on the date when creating the `populatedCoupon`. I think it makes more sense to only do this when picking the expiry date, so I moved this logic to the date picker in dc374d2071a8d9bab37dc26ae3fdc749985c6eb8. Let me know what you think about this - thanks!

pmusolino commented:
> Hi @pmusolino - the issue seems to be caused by the trimming of time on the date when creating the populatedCoupon. I think it makes more sense to only do this when picking the expiry date, so I moved this logic to the date picker in https://github.com/woocommerce/woocommerce-ios/commit/dc374d2071a8d9bab37dc26ae3fdc749985c6eb8. Let me know what you think about this - thanks!

It looks like that with this change, also If I change the date, it will be considered as the same day 💭 looks at the video below for an example of what's happening.

https://user-images.githubusercontent.com/495617/168123329-5966d8da-d636-48af-9cde-2b7a329e90ac.mp4

itsmeichigo committed:
Revert "Move the clearing of time from date to the date picker"

This reverts commit dc374d2071a8d9bab37dc26ae3fdc749985c6eb8.

itsmeichigo committed:
Check expiry date by trimming time

itsmeichigo commented:
Good point! It's not immediately obvious to me why we have a different time for the expiry date - whether it was due to the previous expiry date being set when the store was configured with a different timezone?

However, I suspect since we're trimming time on the updated expiry date, we should trim the time of the initial one as well before comparing. I reverted the previous change and added the trimming in dcd137ef1ea1da21b0c1215e49c2d180ccb6d2af. This works well for different stores, so I hope this is a reliable fix.

Let me know what you think about the solution @pmusolino.

pmusolino committed:
Merge pull request #6848 from woocommerce/issue/6841-fix-remove-expiration-date-when-coupon-is-expired

Edit Coupon: remove expiration date remotely passing an empty string

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:
```diff
@@ -123,6 +123,18 @@ final class AddEditCouponViewModel: ObservableObject {
+        checkAllowedProductsAndCategoriesUpdated(for: coupon) ||
+        checkExpiryDateUpdated(for: coupon) ||
+        checkFreeShippingUpdated(for: coupon) ||
+        checkAllowedProductsAndCategoriesUpdated(for: coupon) ||
```
I think we are checking for `checkAllowedProductsAndCategoriesUpdated` twice.

itsmeichigo replied:
Great catch! Removed in c003d1f9a0e690015b4ab71c2368d02dec324671.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:
```diff
@@ -335,23 +328,63 @@ private extension AddEditCouponViewModel {
+            return false
+        }
+        guard let oldDate = initialCoupon.dateExpires?.startOfDay(timezone: timezone),
+                let newDate = coupon.dateExpires else {
```
Nit: Indentation fix needed.

itsmeichigo replied:
Done here: bd852aeb47afc497e14442871341078561ee5698

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:
```diff
@@ -335,23 +328,63 @@ private extension AddEditCouponViewModel {
+        guard let initialCoupon = self.coupon else {
+            return false
+        }
+        guard let oldDate = initialCoupon.dateExpires?.startOfDay(timezone: timezone),
```
Would it make sense to add a comment to explain why we are trimming time here?

itsmeichigo replied:
Done here: a35e4646f8f2846dcf6805ee2fc18b151c316b3d

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/AddEditCouponViewModelTests.swift:
```diff
@@ -126,7 +126,126 @@ final class AddEditCouponViewModelTests: XCTestCase {
     }
 
-    private enum Localization {
+    func test_hasChangesMade_is_correct_for_coupon_code() {
```
Nice tests. ❤️

selanthiraiyan reviewed (approved):
Nice work @itsmeichigo 🎉 

It tests well. ✅ 

I left a few non-blocker comments.

itsmeichigo committed:
Remove redundant check for allowed products and categories

itsmeichigo committed:
Add comment to clarify the expiry date check

itsmeichigo committed:
Fix indentation
